tools/midenc/src/compiler/emitter.rs

```diff
@@ -0,0 +1,58 @@
+use miden_assembly::ast::{self as masm};
+
+/// The emitter for MASM output.
+
+pub enum MASMAst {
+    Program(masm::ProgramAst),
+    Module(masm::ModuleAst),
+}
+
+pub struct MASMEmitter {
+    writer: BufferWriter,
+}
+
+impl miden_diagnostics::Emitter for MASMEmitter {
+
+    fn buffer(&self) -> Buffer {
+        self.writer.buffer()
+    }
+
+    fn print(&self, buffer: Buffer) -> std::io::Result<()> {
+        self.writer.print(&buffer)
+    }
+}
+
+impl MASMEmitter {
+
+    pub fn new () -> Self {
+        Self {
+            writer: BufferWriter::new(ColorChoice::Auto),
+        }
+    }
+}
+
+impl Pass for Emitter {
+    type Input = MASMAst;
+    type Output = ();
+
+    /// Runs the emitter on the AST 
+    ///
+    /// Errors should be reported via the registered error handler,
+    /// Passes should return `Err` to signal that the pass has failed
+    /// and compilation should be aborted
+    fn run(&mut self, input: Self::Input) -> anyhow::Result<Self::Output> {
+        self.writer.write!("{}", input);
+    }
+
+    /// Implementation of Pass::chain
```

Comment:
You don't have to implement `chain`, as it has a default implementation for all implementations of `Pass`. The inability to meaningfully chain the pass is implicit in the `Output` type of the pass, as a subsequent pass would necessarily have to have an `Input` type of `()`, which is essentially useless.

Writing the generated MASM to a file probably doesn't need to be a pass - though it could still be useful. We will need to have more control over how/where the MASM is written though, so it's probably better to have a dedicated type for the pass which carries all of the options which control that behavior, e.g. `MasmEmitter` or something like that.